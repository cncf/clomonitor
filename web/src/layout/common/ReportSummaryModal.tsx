import { useState } from 'react';
import { FiMoon, FiSun } from 'react-icons/fi';

import CodeBlock from './CodeBlock';
import Modal from './Modal';
import styles from './ReportSummaryModal.module.css';
import Tabs from './Tabs';

interface OpenModalStatus {
  status: boolean;
  name?: string;
}

interface ReportSummaryTheme {
  name: string;
  icon: JSX.Element;
}

const DEFAULT_THEME = 'light';
const THEMES: ReportSummaryTheme[] = [
  {
    name: DEFAULT_THEME,
    icon: <FiSun />,
  },
  { name: 'dark', icon: <FiMoon /> },
];

interface Props {
  orgName: string;
  projectName: string;
  openStatus: OpenModalStatus;
  onCloseModal: () => void;
}

const ReportSummaryModal = (props: Props) => {
  const origin = window.location.origin;
  const [theme, setTheme] = useState<string>(DEFAULT_THEME);
  const image = `${origin}/api/projects/${props.orgName}/${props.projectName}/report-summary?theme=${theme}`;
  const projectLink = `${origin}/projects/${props.orgName}/${props.projectName}`;
  const markdownLink = `[![CLOMonitor report summary](${image})](${projectLink})`;
  const asciiLink = `${projectLink}[image:${image}[CLOMonitor report summary]]`;
  const htmlLink = `<a href="${projectLink}" rel="noopener noreferrer" target="_blank"><img src="${image}" alt="CLOMonitor report summary" /></a>`;

  return (
    <Modal
      header="Embed report summary"
      onClose={props.onCloseModal}
      open={props.openStatus.status && props.openStatus.name === 'reportSummary'}
    >
      <div className="w-100 position-relative">
        <label
          className={`w-100 text-primary text-uppercase fw-bold border-bottom mb-3 ${styles.label}`}
          htmlFor="theme"
        >
          Theme
        </label>
        <div className="d-flex flex-row mb-3">
          {THEMES.map((themeOpt: ReportSummaryTheme) => {
            return (
              <div className="form-check me-4" key={`radio_theme_${themeOpt.name}`}>
                <input
                  className="form-check-input"
                  type="radio"
                  name="theme"
                  id={themeOpt.name}
                  value={themeOpt.name}
                  checked={theme === themeOpt.name}
                  onChange={() => setTheme(themeOpt.name)}
                  required
                  readOnly
                />
                <label className="form-label text-capitalize form-check-label" htmlFor={themeOpt.name}>
                  <div className="d-flex flex-row align-items-center">
                    {themeOpt.icon}
                    <span className="ms-1">{themeOpt.name}</span>
                  </div>
                </label>
              </div>
            );
          })}
        </div>

        <div className="mt-4">
          <label className={`w-100 text-primary text-uppercase fw-bold border-bottom mb-3 ${styles.label}`}>Code</label>
          <Tabs
            className="pt-2"
            tabs={[
              {
                name: 'markdown',
                title: 'Markdown',
                content: (
                  <CodeBlock
                    language="markdown"
                    content={markdownLink}
                    label="Copy report summary markdown link to clipboard"
                  />
                ),
              },

              {
                name: 'ascii',
                title: 'AsciiDoc',
                content: (
                  <CodeBlock
                    language="asciidoc"
                    content={asciiLink}
                    label="Copy report summary Ascii link to clipboard"
                  />
                ),
              },
              {
                name: 'html',
                title: 'HTML',
                content: (
                  <CodeBlock language="html" content={htmlLink} label="Copy report summary html link to clipboard" />
                ),
              },
            ]}
            active="markdown"
            noDataContent="Sorry, the information for this is missing."
          />
        </div>

        <div className="mt-4 d-flex flex-column">
          <label className={`text-primary text-uppercase fw-bold border-bottom mb-4 ${styles.label}`}>Preview</label>

          <div className={`mx-auto my-3 ${styles.imgWrapper}`}>
            <img src={image} alt="CLOMonitor report summary" />
          </div>
        </div>
      </div>
    </Modal>
  );
};

export default ReportSummaryModal;
