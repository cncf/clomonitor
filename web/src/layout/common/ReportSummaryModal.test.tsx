import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ReportSummaryModal from './ReportSummaryModal';

const mockOnCloseModal = jest.fn();

const defaultProps = {
  orgName: 'org',
  projectName: 'proj',
  openStatus: { status: true, name: 'reportSummary' },
  onCloseModal: mockOnCloseModal,
};

describe('ReportSummaryModal', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<ReportSummaryModal {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders proper content', () => {
      render(<ReportSummaryModal {...defaultProps} />);

      expect(screen.getByText('Embed report summary')).toBeInTheDocument();
      expect(screen.getByText('Theme')).toBeInTheDocument();
      expect(screen.getByRole('radio', { name: 'light' })).toBeInTheDocument();
      expect(screen.getByRole('radio', { name: 'light' })).toBeChecked();
      expect(screen.getByRole('radio', { name: 'dark' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open tab markdown' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open tab ascii' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open tab html' })).toBeInTheDocument();
      expect(screen.getByText('Preview')).toBeInTheDocument();
      expect(screen.getByAltText('CLOMonitor report summary')).toBeInTheDocument();
    });

    it('renders markdown tab', () => {
      render(<ReportSummaryModal {...defaultProps} />);

      expect(screen.getByText('Embed report summary')).toBeInTheDocument();
      expect(screen.getAllByText('Markdown')).toHaveLength(2);

      expect(screen.getByTestId('code')).toHaveTextContent(
        '[![CLOMonitor report summary](http://localhost/api/projects/org/proj/report-summary?theme=light)](http://localhost/projects/org/proj)'
      );
      expect(
        screen.getByRole('button', { name: 'Copy report summary markdown link to clipboard' })
      ).toBeInTheDocument();
    });

    it('renders ascii tab', () => {
      render(<ReportSummaryModal {...defaultProps} />);

      expect(screen.getAllByText('AsciiDoc')).toHaveLength(2);
      const btn = screen.getByRole('button', { name: 'Open tab ascii' });
      expect(btn).toHaveTextContent('AsciiDoc');
      userEvent.click(btn);

      expect(screen.getByTestId('code')).toHaveTextContent(
        'http://localhost/projects/org/proj[image:http://localhost/api/projects/org/proj/report-summary?theme=light[CLOMonitor report summary]]'
      );
      expect(screen.getByRole('button', { name: 'Copy report summary Ascii link to clipboard' })).toBeInTheDocument();
    });

    it('renders html tab', () => {
      render(<ReportSummaryModal {...defaultProps} />);

      expect(screen.getAllByText('HTML')).toHaveLength(2);
      const btn = screen.getByRole('button', { name: 'Open tab html' });
      expect(btn).toHaveTextContent('HTML');
      userEvent.click(btn);

      expect(screen.getByTestId('code')).toHaveTextContent(
        '<a href="http://localhost/projects/org/proj" rel="noopener noreferrer" target="_blank"><img src="http://localhost/api/projects/org/proj/report-summary?theme=light" alt="CLOMonitor report summary" /></a>'
      );
      expect(screen.getByRole('button', { name: 'Copy report summary html link to clipboard' })).toBeInTheDocument();
    });
  });
});
