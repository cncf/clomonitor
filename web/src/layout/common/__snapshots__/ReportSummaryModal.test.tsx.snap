// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportSummaryModal creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="modal-backdrop activeBackdrop"
      data-testid="modalBackdrop"
    />
    <div
      aria-modal="true"
      class="modal modal active d-block"
      role="dialog"
    >
      <div
        class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
      >
        <div
          class="modal-content rounded-0 border border-3 mx-auto content"
        >
          <div
            class="modal-header rounded-0 d-flex flex-row align-items-center header"
          >
            <div
              class="modal-title h5 m-2 flex-grow-1"
            >
              Embed report summary
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="modal-body p-4 h-100 d-flex flex-column"
          >
            <div
              class="w-100 position-relative"
            >
              <label
                class="w-100 text-primary text-uppercase fw-bold border-bottom mb-3 label"
                for="theme"
              >
                Theme
              </label>
              <div
                class="d-flex flex-row mb-3"
              >
                <div
                  class="form-check me-4"
                >
                  <input
                    checked=""
                    class="form-check-input"
                    id="light"
                    name="theme"
                    readonly=""
                    required=""
                    type="radio"
                    value="light"
                  />
                  <label
                    class="form-label text-capitalize form-check-label"
                    for="light"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          cx="12"
                          cy="12"
                          r="5"
                        />
                        <line
                          x1="12"
                          x2="12"
                          y1="1"
                          y2="3"
                        />
                        <line
                          x1="12"
                          x2="12"
                          y1="21"
                          y2="23"
                        />
                        <line
                          x1="4.22"
                          x2="5.64"
                          y1="4.22"
                          y2="5.64"
                        />
                        <line
                          x1="18.36"
                          x2="19.78"
                          y1="18.36"
                          y2="19.78"
                        />
                        <line
                          x1="1"
                          x2="3"
                          y1="12"
                          y2="12"
                        />
                        <line
                          x1="21"
                          x2="23"
                          y1="12"
                          y2="12"
                        />
                        <line
                          x1="4.22"
                          x2="5.64"
                          y1="19.78"
                          y2="18.36"
                        />
                        <line
                          x1="18.36"
                          x2="19.78"
                          y1="5.64"
                          y2="4.22"
                        />
                      </svg>
                      <span
                        class="ms-1"
                      >
                        light
                      </span>
                    </div>
                  </label>
                </div>
                <div
                  class="form-check me-4"
                >
                  <input
                    class="form-check-input"
                    id="dark"
                    name="theme"
                    readonly=""
                    required=""
                    type="radio"
                    value="dark"
                  />
                  <label
                    class="form-label text-capitalize form-check-label"
                    for="dark"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        fill="none"
                        height="1em"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                        />
                      </svg>
                      <span
                        class="ms-1"
                      >
                        dark
                      </span>
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="mt-4"
              >
                <label
                  class="w-100 text-primary text-uppercase fw-bold border-bottom mb-3 label"
                >
                  Code
                </label>
                <div
                  class="pt-2"
                >
                  <ul
                    class="nav nav-tabs tabs"
                  >
                    <li
                      class="nav-item"
                    >
                      <button
                        aria-label="Open tab markdown"
                        class="btn nav-item rounded-0 lightText btn active btn-secondary active"
                      >
                        <span
                          class="d-none d-sm-block"
                        >
                          Markdown
                        </span>
                        <span
                          class="d-block d-sm-none"
                        >
                          Markdown
                        </span>
                      </button>
                    </li>
                    <li
                      class="nav-item"
                    >
                      <button
                        aria-label="Open tab ascii"
                        class="btn nav-item rounded-0 lightText btn"
                      >
                        <span
                          class="d-none d-sm-block"
                        >
                          AsciiDoc
                        </span>
                        <span
                          class="d-block d-sm-none"
                        >
                          AsciiDoc
                        </span>
                      </button>
                    </li>
                    <li
                      class="nav-item"
                    >
                      <button
                        aria-label="Open tab html"
                        class="btn nav-item rounded-0 lightText btn"
                      >
                        <span
                          class="d-none d-sm-block"
                        >
                          HTML
                        </span>
                        <span
                          class="d-block d-sm-none"
                        >
                          HTML
                        </span>
                      </button>
                    </li>
                  </ul>
                </div>
                <div
                  class="tab-content mt-4"
                >
                  <div
                    class="tab-pane fade show active"
                  >
                    <div
                      class="d-flex flex-row align-items-center pb-2 codeBlock"
                      data-testid="code"
                    >
                      <pre
                        style="display: block; overflow-x: auto; padding: 1rem 0.5rem; background: rgb(248, 248, 255); margin-bottom: 0px;"
                      >
                        <code
                          class="language-markdown"
                          style="white-space: pre;"
                        >
                          <span>
                            [
                          </span>
                          <span
                            style="color: rgb(33, 145, 97);"
                          >
                            ![CLOMonitor report summary
                          </span>
                          <span>
                            ](
                          </span>
                          <span
                            style="color: rgb(187, 102, 136);"
                          >
                            http://localhost/api/projects/org/proj/report-summary?theme=light
                          </span>
                          <span>
                            )](
                          </span>
                          <span
                            style="color: rgb(187, 102, 136);"
                          >
                            http://localhost/projects/org/proj
                          </span>
                          <span>
                            )
                          </span>
                        </code>
                      </pre>
                      <div
                        class="position-relative ms-3"
                      >
                        <button
                          aria-label="Copy report summary markdown link to clipboard"
                          class="btn btn-sm btn-primary rounded-circle btn"
                          type="button"
                        >
                          <div
                            aria-hidden="true"
                            class="d-flex flex-row align-items-center"
                          >
                            <svg
                              fill="none"
                              height="1em"
                              stroke="currentColor"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="2"
                              viewBox="0 0 24 24"
                              width="1em"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <rect
                                height="13"
                                rx="2"
                                ry="2"
                                width="13"
                                x="9"
                                y="9"
                              />
                              <path
                                d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
                              />
                            </svg>
                          </div>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="mt-4 d-flex flex-column"
              >
                <label
                  class="text-primary text-uppercase fw-bold border-bottom mb-4 label"
                >
                  Preview
                </label>
                <div
                  class="mx-auto my-3 imgWrapper"
                >
                  <img
                    alt="CLOMonitor report summary"
                    src="http://localhost/api/projects/org/proj/report-summary?theme=light"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="modal-footer p-3"
          >
            <button
              aria-label="Close modal"
              class="btn btn-sm rounded-0 btn-secondary text-uppercase"
              type="button"
            >
              <div
                class="d-flex flex-row align-items-center"
              >
                <svg
                  class="me-2"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <div>
                  Close
                </div>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
